app/services/flight_data_service.rb

```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+class FlightDataService
+  def initialize(arg, fare_category)
+    @arg = arg
+    @fare_category = fare_category
+  end
+
+  def self.execute(arg, fare_category)
+    new(arg, fare_category).flight_datas
+  end
+
+  def flight_datas
+    arg.map do |itinerary|
+      build_data(itinerary)
+    end
+  end
+
+  private
+
+  attr_reader :arg, :fare_category
+
+  def build_data(obj)
+    flight = Flight.create(fare_category: fare_category.presence || 'economic')
+    build_price(flight.id, obj[:price])
+    itineraries(flight.id, obj[:legs])
+    flight
+  end
+
+  def build_price(flight_id, obj)
+    Price.create({
+      flight_id: flight_id,
+      air_miles: obj[:air_miles],
+      currency: obj[:currency],
+      value: obj[:raw],
+      formatted_price: obj[:formatted]
+    })
+  end
+
+  def itineraries(flight_id, legs)
+    legs.each do |leg|
+      flight_detail = build_flight_detail(leg)
+      RelatedConnection.find_or_create_by!(
+        flight_id: flight_id,
+        flight_detail_id: flight_detail.id
+      )
+      build_connections(leg[:segments], flight_id, flight_detail.id) if (leg[:stopCount]).positive?
+    end
+  end
+
+  def build_flight_detail(obj)
+    flight_detail = FlightDetail.find_or_create_by!({
+      origin: obj.dig(:origin, :name),
+      destiny: obj.dig(:destination, :name),
+      origin_airport: obj.dig(:origin, :displayCode),
+      destination_airport: obj.dig(:destination, :displayCode),
+      departure_time: obj[:departure],
+      arrival_time: obj[:arrival]
+    })
+    number_and_airline(flight_detail, obj)
+  end
+
+  def number_and_airline(flight_detail, obj)
+    if one_way_trip?(obj)
+      flight_detail.flight_number = obj.dig(:segments, 0, :flightNumber)
+      flight_detail.name_airline = obj.dig(:segments, 0, :operatingCarrier, :name)
+    elsif !segments?(obj)
+      flight_detail.flight_number = obj[:flightNumber]
+      flight_detail.name_airline = obj.dig(:operatingCarrier, :name)
+    end
+    flight_detail.save
+    flight_detail
+  end
```

Comment:
Imagino que isso poderia ser bem mais simples, o que você acha de algo assim:
```suggestion
    def build_flight_detail(obj)
    node = choose_node_for_number_and_airline_name

    FlightDetail.find_or_create_by!({
      origin: obj.dig(:origin, :name),
      destiny: obj.dig(:destination, :name),
      origin_airport: obj.dig(:origin, :displayCode),
      destination_airport: obj.dig(:destination, :displayCode),
      departure_time: obj[:departure],
      arrival_time: obj[:arrival],
      flight_number: node.dig(:flightNumber),
      name_airline: node.dig(:operatingCarrier, :name)
    })
  end

  def choose_node_for_number_and_airline_name(obj)
    return obj.dig(:segments, 0) if one_way_trip?(obj)
    return obj if !segments(obj)
  end
```

Ao rodar todos os testes eles passaram, portanto teoricamente o código acima deveria funcionar.

Reply:
Com pequenas modificações o código funcionou sim :wink:
Ajustado em: https://github.com/Laryssa-lfa/flight-company/pull/5/commits/2bbfd36dc212fd27d0036583c6e9dbf230f4a7b0.